test/compat/render-chunked.test.js

```diff
@@ -66,4 +68,85 @@ describe('renderToChunks', () => {
 			'</div>'
 		]);
 	});
+
+	it('should encounter no circular references when rendering a suspense boundary subtree', async () => {
+		const { Suspender, suspended } = createSuspender();
+
+		const visited = new Set();
+		let circular = false;
+
+		function CircularReferenceCheck() {
+			let root = this[VNODE];
+			while (root !== null && root[PARENT] !== null) {
+				if (visited.has(root)) {
+					// Can't throw an error here, _catchError handler will also loop infinitely
+					circular = true;
+					break;
+				}
+				visited.add(root);
+				root = root[PARENT];
+			}
+			return <p>it works</p>;
+		}
+
+		const result = [];
+		const promise = renderToChunks(
+			<div>
+				<Suspense fallback="loading...">
+					<Suspender>
+						<CircularReferenceCheck />
+					</Suspender>
+				</Suspense>
+			</div>,
+			{ onWrite: (s) => result.push(s) }
+		);
+
+		suspended.resolve();
+		await promise;
+
+		if (circular) {
+			throw new Error('CircularReference');
+		}
+
+		expect(result).to.deep.equal([
+			'<div><!--preact-island:16-->loading...<!--/preact-island:16--></div>',
+			'<div hidden>',
+			createInitScript(1),
+			createSubtree('16', '<p>it works</p>'),
+			'</div>'
+		]);
+	});
+
+	it('should support using useId hooks inside a suspense boundary', async () => {
+		const { Suspender, suspended } = createSuspender();
```

Comment:
You can add children here, so if we do the following twice, then you can actually assert the behaviour of your issue where you don't want two adjacent trees to have the same id

```suggestion
		const { Suspender, suspended } = createSuspender(<ComponentWithId />);
```

Reply:
`createSuspender` doesn't take any arguments?
But changing it to
```jsx
				<ComponentWithId />
				<Suspense fallback="loading...">
					<Suspender>
						<ComponentWithId />
					</Suspender>
					<Suspender>
						<ComponentWithId />
					</Suspender>
				</Suspense>
```
could work, then the expected result is P0-0, P0-0, P0-1

Reply:
It does on main

Reply:
am I missing something here? https://github.com/preactjs/preact-render-to-string/blob/main/test/utils.jsx#L18

Reply:
If you want to finish it up, that sounds good to me!